src/Processors/Transforms/MergingAggregatedMemoryEfficientTransform.h

```diff
@@ -82,15 +84,14 @@ class GroupingAggregatedTransform final : public IProcessor
     bool has_two_level = false;
 
     bool all_inputs_finished = false;
-    bool read_from_all_inputs = false;
-    std::vector<bool> read_from_input;
+    bool initialized_index_to_input = false;
+    std::vector<InputPorts::iterator> index_to_input;
+    HashSet<uint64_t> wait_input_ports_numbers;
```

Comment:
The key optimization here is using the HashSet to track which input ports still need data instead of scanning all inputs each time.

In the original code, it was looping through all inputs every time to check if each one needed more data for the current bucket, which is O(n) for n inputs.

With the HashSet, it only needs to check the inputs that were previously marked as needing data. As it gets chunks from those waiting inputs, it removes them from the HashSet.

This avoids repeatedly scanning inputs that don't need more data for the current bucket. The contains, insert, and erase operations on the HashSet are O(1) average complexity rather than O(n) for a linear scan of inputs.